### ISSUE TITLE
Converting RCS configs over to Engine Globals

### ISSUE THREAD

**Issue by NathanKell** [KSP-RO/RealismOverhaul#1748]
Done, except for Coatl Aerospace. Didn't know the nozzle multiplier to use there. @PhineasFreak  you did that one right? Please replace the RCS cost mults with the right nozzle counts (%RcsNozzles = 2 for, say, 2 nozzles per part).

Also there might be non-4-nozzle RCS that have a generic-RCS config but didn't have a cost modifier, in which case they also need to have nozzle number specified.

**Comment by PhineasFreak** [comment]
Yes, i will fix them and push the changes to the dev branch as soon as i can.

**Comment by NathanKell** [comment]
Awesome!
To support them I also added a couple fields to the engine variants code. Think they're all working fine. I need to go back and apply them to the LR87...

**Comment by PhineasFreak** [comment]
BTW, is it possible for a part config to have multiple ModuleEngineConfigs defined? There is a part (https://github.com/KSP-RO/RealismOverhaul/blob/master/GameData/RealismOverhaul/RO_SuggestedMods/CoatlAerospace/RO_CA_ProbesPlus_Control.cfg#L566-L646) that is both a high powered RCS module and a generic engine (like the 1 kN thruster). With the previous RCS setup you could only define a propellant combination for the RCS part but not the engine part.

**Comment by PhineasFreak** [comment]
OK, the PR is up: https://github.com/KSP-RO/RealismOverhaul/pull/1750

**Comment by raidernick** [comment]
Can someone explain to me why you'd ever want to do this. I can understand the engine configs because it's affected by Rp0 and testflight and using globals helps there. RCS has nothing to do with tech tree placement and isn't affected by testflight at all. I can't see any reason to ever waste time doing this over just modifying the fuel types used in the config this adds more extra steps than just doing it the old way...

**Comment by PhineasFreak** [comment]
@raidernick The RCS thrusters do have TF compatibility: https://github.com/KSP-RO/RealismOverhaul/blob/d97a1dab4a13230e02806746a60fa5f29b4b5288/GameData/RealismOverhaul/Engine_Configs/RCS_Config.cfg#L199-L244.

Besides, RCS parts were the first iteration of the global engine config. Also, it is easier to set two global dields and be done with the part than copying and pasting the same thing over and over.

**Comment by raidernick** [comment]
I'm sorry but I just can't see how this would be useful in 99% of cases. I only seeing this to be valuable to some generic "all uses" thrusters like the stock ones. Any thruster that has one specific purpose for use on one specific spacecraft only, this is simply adding extra steps for no reason other than to add extra steps while missing the ability to give exact values.

From what I see so far there is no reason I would ever even consider using these unless I was making a one size fits all thruster like the stock ones, which I will likely never do. I need a better reason to do it other than just testflight. The generic configs give me a good reason to use them, what I see here gives no good reason at all.

**Comment by PhineasFreak** [comment]
So you prefer to clone the same part over and over to add to it different propellant combinations? Most of the mod thrusters can be considered generic ones (i could add an exception for the SSTU ones, especially the Centaur-alike thrusters).

IIRC the **generic** RCS thrusters were always set to have a wide range of possible propellant options. If you want to set one to have a specific function (like the stock ATV RCS thruster - uses only MMH and NTO) then you are free to do so.

**Comment by raidernick** [comment]
Why would I clone a probe over and over for different propellant combinations when it only uses one combination of propellant? If someone wants to cheat and use different fuels in a probe that didn't use those fuels, the burden of changing the configs should lie on them not me. This is supposed to be REALISM overhaul, not use whatever fuel you like in something that was not built for it overhaul. It's like asking why the rl10 engine configs don't support the usage of udmh and nto as fuel; it's because they were not designed for that and if a player wants it to use that they need to edit it themselves.

And yes let's talk about if it had a specific propellant like the atv rcs thruster. Now instead of adding a simple change to the RO configs for it like:
`@MODULE[ModuleRCS*]
{
@PROPELLANT[Monopropellant]
		{
			name = UDMH
			ratio = 0.357
		}
		PROPELLANT
		{
			name = NTO
			ratio = 0.643
		}
}`

I now need to use a generic config in the RO config, then write ANOTHER config for the same part that runs after RO-RCS to remove the configs I don't want on the part and overwrite the thruster power and isp to be the EXACT published values for that specific spacecraft. This is like the generic engine configs pass that overwrites for[realismoverhaul] and if you want to actually make changes to the engine you need to write a second config for every rcs part to run after the ro-rcs pass just like RealismOverhaulEngines. This is double the work for no payoff at all. So yes, yes I would rather copy and paste that simple bit of code above for _separate_ and totally _unrelated_ spacecraft with built in rcs thrusters.

**Comment by raidernick** [comment]
The problem here is that an rl10 is an rl10. Every specific TYPE of that rl10 is EXACTLY the same. For isntance all Rl10-B2s are the same. You are grouping rcs thrust configs here by the fuel types. Not every single rcs thruster that uses udmh and nto is the same. In fact, most rcs thrusters are built specifically for the spacecraft they are used for. A generic "catch all" in this case is worthless.

**Comment by PhineasFreak** [comment]
OK, then let's lock all generic parts to have specific applications...

**Comment by raidernick** [comment]
You clearly aren't reading what I'm writing at all. I'm saying any rcs module that's part of it's parent spacecraft model, this system is not useful for. I literally said it's only applicable to generic rcs port parts. For any one specific model that has built in thrusters or a thruster block modeled specifically to fit onto only one type of spacecraft, using generic configs for them is adding more work as you then need to remove all the configs for fuels that it doesn't use and overwrite the mix rates, thruster power and even isp values for the parts.

The entire point of this was asking why in the case of spacecraft like mine that have combined parent models and rcs transforms that I should be pressured into using this system when unlike the engine configs system, it has no value on these specific types of parts.

**Comment by PhineasFreak** [comment]
Who ever said that you are forced to follow this convention? A good example of this are the global engine configs: you can set your own (as you have done and maintain them yourself) or use the "community-driven" one.

If you don't set the **useRcsConfig** line in the RO part patch then nothing will ever touch it, as it was done from the beginning. Why is this an issue **now**?

**Comment by NathanKell** [comment]
I wasn't aware you used generic-RCS-configs, @raidernick, because my understanding is that you don't have generic RCS; indeed, as you say, you have specific RCS. If that specific RCS exists on more than one part, then that should get any engine global config so only one thing need be maintained (and so other things implementing that thruster can have the same config), but that's entirely separate from changes to the generic-RCS system.

**Comment by raidernick** [comment]
Because I AM being pressured into using the generic configs because apparently not using them breaks the rp0 tree or causes problems with testflight or some crap. I don't know because I don't use either or those but I got complaints about it lately and @NathanKell wants me to change them to use generic ones for the r-7 engines which is a MAJOR project as every single engine and tank part needs to be changed then about 40 craft files need to be rebuilt from scratch(and that's just for my mod not even kk or fasa which I will need to redo also). 

Now I feel that because of the same damn mod I don't like or use, and every part being converted to this I am being pressured to change all that crap also. And unlike the generic configs for engine, the rcs ones have no actual value to my parts, kk parts or fasa parts all of which I maintain now would also need to be changed.

**Comment by NathanKell** [comment]
You are being asked to use the *engine* configs. For the reasons we discussed. The generic RCS configs are not that. As I ltierally just said, if you have multiple parts with the exact same thruster (for example, the same Isayev thruster with the same performance), *that's* when using a global config would make sense.

That has nothing to do with *generic* RCS.

**Comment by raidernick** [comment]
@NathanKell your comment didn't show up until after I was done typing that and hit post

**Comment by NathanKell** [comment]
Ah sorry! :]

**Comment by raidernick** [comment]
@NathanKell My original question was a simple question asking the value this had to parts that had one specific type of thruster and only that one type and if there was a reason I should be using it. You answered my question exactly for what I was asking about. 

@PhineasFreak I don't understand what he was trying to say that I was copying my parts over and over to allow multiple configs or something when it had no reason to use them. That's why I was confused about it as it had nothing to do with what I was trying to find out.

**Comment by NathanKell** [comment]
Ah, cool. Glad it's sorted out! :)

I would say that if you have parts that use the same exact thruster performance, it might be worth using global configs there, if only because (a) for TF reasons and (b) if you later find something showing different performance, you'd only have to make the change in one place.

And yes, I know the TF config system isn't currently hooked up right for RCS. That's not working as desired and is on my todo list. :)

**Comment by raidernick** [comment]
@NathanKell I can only think of one instance where the rcs configs for more than one part is the same and I believe it's with the soyuz/zond/lok descent modules. I don't believe there are any published values for lok or zond so I used the soyuz ones for it. But I'm pretty sure in every other case, since I mostly do probes, that the rcs is different on a case by case basis. These spacecraft generally have the rcs built specifically for that probe only(sometimes they might take an existing thruster, but it's generally modified which changes its stats) so the stats are always different.

**Comment by NathanKell** [comment]
Ya, makes sense!

**Comment by pap1723** [comment]
Any update on this?
